src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -53,6 +53,30 @@ typedef struct gavl_tree {
 
 #define STACK_EMPTY(stack) (!stack##_sp)
 
+#if MPL_THREAD_PACKAGE_NAME != MPL_THREAD_PACKAGE_NONE
+
+#define GAVL_THREAD_CS_ENTER                                                              \
+    do {                                                                                  \
+        if (gavl_tree_iptr->need_thread_safety) {                                         \
```

Comment:
I guess the question is how is `need_thread_safety` set? If it is set by MPICH, then we probably can remove it altogether and require MPICH place critical section where needed.

Reply:
It is set by MPICH, as far as I can tell. Thread-safety probably can move to MPICH without much issue. Does anyone who worked on it have an issue with that?

Reply:
I have similar feeling. It might be simpler (i.e., we do not need check `need_thread_safety` given by the user) if we move the critical section to MPICH. The GAVL routines can be simply thread unsafe.

Reply:
I prefer doing it in MPICH, but @KaimingOuyang was adamant about doing it in MPL.  I'll let him respond.

Reply:
Actually, I have no problem to let MPICH provide thread-safety. I did it in MPL because I don't want to complicate the upper layer or say users to think about the thread-safety issue when using gavl tree.

Reply:
@KaimingOuyang  That was one of my review comments.  You insisted very strongly on putting it in MPL.  I gave up after arguing for a while.  :-)

Reply:
Alright, maybe I was not persuaded ;)
I can delete this parameter and related snippets in my PR.

Reply:
> Actually, I have no problem to let MPICH provide thread-safety. I did it in MPL because I don't want to complicate the upper layer or say users to think about the thread-safety issue when using gavl tree.

Hi Kaiming, what you probably didn't realize is that, the user, MPICH in this case, have to think about the thread-safety issue on every thing. It is a critical piece in MPICH's performance and it needs to be managed by MPICH explicitly.

Reply:
@hzhou Maybe you are right. However, probably this is not a good example, I actually

Reply:
>  If we decide to put all thread-safety consideration to MPICH, I can definitely move it to MPICH since this convention should be consistent.

Yes. There is set of (overly complicated) infrastructure in MPICH to control critical sections. Having something separate in MPL outside MPICH makes it more complicated.

MPL memory trace is a debugging tool, which is an outlier.

Reply:
okay, sounds good.